Update incomplete, partially unknown and broken stubs

Avasam opened the pull request:
First of, thanks for this fork, it's been easy to use, in parts thanks to the stubs. However, there's a few issues with those:

A handful of improvements to make the stubs complete, more usable and follow up-to-date stub typing standards:
- Explicitely re-export imports meant to be re-exported
- Now passes mypy, pyright and Flake8-PYI
- Explicitely return `None`. Otherwise mypy infers as `Any` and pyright as `Unknown`
- Mark type-only classes with `type_check_only`. Those classes are type references and cannot be imported at runtime.
- Prefer `|` rather than `Union` or `Optional`
- Prefer `collections.abc.Callable` over `typing.Callable`
- Fixed the syntax error at `memorybreakpoint`'s return type
- Explicitely type `Position` as a `TypeAlias`
- Fixed the type of optional parameters that have a default value of `None`

Avasam commented on python-stubs/dolphin/__init__.pyi:
```diff
@@ -1 +1 @@
-from . import controller, event, gui, memory, savestate
+from . import controller as controller, event as event, gui as gui, memory as memory, savestate as savestate
```
In the latest release, `dolphin` isn't a module, I assume it is now a module in master? Otherwise this `__init__` would be wrong.

Avasam commented on python-stubs/dolphin/controller.pyi:
```diff
@@ -4,9 +4,10 @@ Module for programmatic inputs.
+from typing import TypedDict, type_check_only
 
 
+@type_check_only
```
If you could, these would be beneficial to have during runtime, otherwise you have to use them as such to use them as type annotations:
![image](https://user-images.githubusercontent.com/1350584/222532310-c4d8320c-3ce5-402c-976b-0d6c55a9d083.png)

See below, used directly is fine, as it is inferred as the expected parameter type, but inferred as `dict[str, bool]` is not:
![image](https://user-images.githubusercontent.com/1350584/222533039-f3810781-0c09-40bc-83ad-3b106ee39cd1.png)

Felk commented:
Hey, thanks for all the work! Very appreciated!

I'll try to take a deep look over the weekend.

Felk reviewed (approved):
I finally took a deeper look (sorry, that was a _long_ weekend) and your changes look good :) regarding using collections.abc.Callable instead of typing.Callable I think this also suggests I should update to Python 3.9 or up, correct? Because looks like the transition from the former to the latter seems to be a python 3.9 thing

Avasam commented:
> regarding using collections.abc.Callable instead of typing.Callable I think this also suggests I should update to Python 3.9 or up, correct? Because looks like the transition from the former to the latter seems to be a python 3.9 thing

No need. You can (and should) use the latest type features in stub files because they're not used at runtime.

Felk commented:
Ah, I see. Well I updated to Python 3.11 today just in case anyway. I
suppose that means I can change some imports from "typing_extensions" to
just "typing" now too, right?

On Sun, 17 Sept 2023, 18:14 Avasam, ***@***.***> wrote:

> regarding using collections.abc.Callable instead of typing.Callable I
> think this also suggests I should update to Python 3.9 or up, correct?
> Because looks like the transition from the former to the latter seems to be
> a python 3.9 thing
>
> No need. You can (and should) use the latest type features in stub files
> because they're not used at runtime.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Felk/dolphin/pull/34#issuecomment-1722510644>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPHK2WDRZIAIL33S5PW2ULX24OWPANCNFSM6AAAAAAVNY7ZYQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

Felk commented:
I added the typed dicts to the runtime and removed the `@type_check_only` annotations in 16036d7628249aaf67167f6ec04972f83a7d825b, hopefully that was correct the way I did it

Avasam commented:
> I added the typed dicts to the runtime and removed the `@type_check_only` annotations in [16036d7](https://github.com/Felk/dolphin/commit/16036d7628249aaf67167f6ec04972f83a7d825b), hopefully that was correct the way I did it

Untested but looks correct at a glance
